_posts/2023-03-27-group-equivariant-cnn-2/group-equivariant-cnn-2.Rmd

```diff
@@ -0,0 +1,374 @@
+```
+
+```{r, echo=FALSE, eval=TRUE, fig.alt = "Same goat, rotated up by 90 degrees."}
+knitr::include_graphics("images/z2.jpg")
+```
+
+## Step 2: The lifting convolution
+
+We want to make use of existing, efficient `torch` functionality as much as possible. Concretely, we want to use `nn_conv2d()`. What we need, though, is a convolution kernel that's equivariant not just to translation, but also, to the action of $C_4$. This can be achieved by having one kernel each for every possible rotation.
+
+Implement that idea is exactly what `LiftingConvolution` does. The principle is the same as before: First, the grid is rotated, and then, the kernel (weight matrix) is re-sampled to the transformed grid.
+
+Why, though, call this a *lifting convolution*? The usual convolution kernel operates on $\mathbb{R}^2$; our extended version operates on combinations of $\mathbb{R}^2$ and $C_4$. In math speak, it has been *lifted* to the [semi-direct product](https://en.wikipedia.org/wiki/Semidirect_product) $\mathbb{R}^2\rtimes C_4$.
+
+```{r}
+lifting_conv <- LiftingConvolution(
+    group = CyclicGroup(order = 4),
+    kernel_size = 5,
+    in_channels = 3,
+    out_channels = 8
+  )
+
+x <- torch::torch_randn(c(2, 3, 32, 32))
+y <- lifting_conv(x)
+y$shape
+```
+
+    [1]  2  8  4 28 28
+
+Since, internally, `LiftingConvolution` uses an additional dimension to realize the product of translations and rotations, the output is not four-, but five-dimensional.
+
+## Step 3: Group convolutions
+
+Now that we're in "group-extended space", we can chain a number of layers where both input and output are *group convolution* layers. For example:
+
+```{r}
+group_conv <- GroupConvolution(
+  group = CyclicGroup(order = 4),
+    kernel_size = 5,
+    in_channels = 8,
+    out_channels = 16
+)
+
+z <- group_conv(y)
+z$shape
+```
+
+    [1]  2 16  4 24 24
+
+All that remains to be done is package this up. That's what `gcnn::GroupEquivariantCNN()` does.
+
+## Step 4: Group-equivariant CNN
+
+We can call `GroupEquivariantCNN()` like so.
+
+```{r}
+cnn <- GroupEquivariantCNN(
+    group = CyclicGroup(order = 4),
+    kernel_size = 5,
+    in_channels = 1,
+    out_channels = 1,
+    num_hidden = 2, # number of group convolutions
+    hidden_channels = 16 # number of channels per group conv layer
+)
+
+img <- torch::torch_randn(c(4, 1, 32, 32))
+cnn(img)$shape
+```
+
+    [1] 4 1
+
+At casual glance, this `GroupEquivariantCNN` looks like any old CNN ... weren't it for the `group` argument.
+
+When now we inspect its output, we see that the additional dimension is gone. That's because after a sequence of group-to-group convolution layers, the module projects down to a representation that, for each batch item, retains channels only. It thus averages not just over locations -- as we normally do -- but over the group dimension as well. A final linear layer will then provide the requested classifier output (of dimension `out_channels`).
```

Comment:
```suggestion
Now, when we inspect its output, we see that the additional dimension is gone. That's because after a sequence of group-to-group convolution layers, the module projects down to a representation that, for each batch item, retains channels only. It thus averages not just over locations -- as we normally do -- but over the group dimension as well. A final linear layer will then provide the requested classifier output (of dimension `out_channels`).
```